### DISCUSSION TITLE
[JIT] Not working with linked dependency

### DISCUSSION THREAD

**Discussion by tonytoth** [tailwindlabs/tailwindcss#7717] [Help]
**What version of Tailwind CSS are you using?**

v3.0.23

**What build tool (or framework if it abstracts the build tool) are you using?**

CRA, Storybook, Rollup

**What version of Node.js are you using?**

v16.13.1

**What browser are you using?**

Chrome

**What operating system are you using?**

macOS

**Reproduction URL**

https://github.com/tonytoth/tailwind-yarn-link

**Describe your issue**

So basically I have a CRA with storybook where I need to test my package which is based on Tailwind CSS and I'm expecting to have the JIT feature available so If I modify the Accordion component, the JIT will be triggered and my Storybook will be updated.
It works if I use my `dist` folder to be watched but it does not work if I'm using `node_modules` where my package is linked.

See `tailwind.config.js`: https://github.com/tonytoth/tailwind-yarn-link/blob/master/example/tailwind.config.js
See `package.json` linking: https://github.com/tonytoth/tailwind-yarn-link/blob/master/example/package.json#L16

**Comment by adamwathan** [comment]
Hey! Converting this to a discussion to see if we can help you figure out the problem first before escalating to an actual bug 👍🏻

**Comment by thecrypticace** [comment]
I tried setting up the reproduction but I couldn't get it to work. The example folder is missing `@storybook/react` but even after installing that and getting the scripts to run I'm running into errors. I don't know enough about storybook to get past the error though.

<img width="847" alt="Screen Shot 2022-03-01 at 13 06 50" src="https://user-images.githubusercontent.com/614993/156224482-0f3ac3ab-246a-4eaa-9237-a21491113416.png">

**Reply by tonytoth** [reply]
Can you please try to refresh the page when you get that error ? (I think it happened to me once)
I will try to fix it tomorrow.

**Reply by tonytoth** [reply]
I managed to fix it, pushed to master.

**Comment by tonytoth** [comment]
I tried to understand what's going on I am not sure if it is a Tailwind issue. From what I see the way Tailwind transfer data for HMR to postCSS is fine. But I don't really understand exactly postCSS and Tailwind with postCSS so I would really appreciate if you guys cand take a look. @adamwathan @thecrypticace

**Reply by thecrypticace** [reply]
I'll be taking a look today! Sorry for the delay.

**Reply by thecrypticace** [reply]
Okay, so the issue here is:
1. Content has the path to a symlink
2. This symlink is otherwise unknown to webpack (or rather it's something it doesn't have to care about)
3. fast-glob (if using a glob; in this case it's not an issue) returns paths relative to the symlink and not the real path of the files
4. We register that symlink-including-path as a dependency which is then not picked up by webpack as something that has changed. The only thing that does is the dist file. My guess is because it resolves those symlinks internally before registering dependency trees. Not 100% certain on that.

There is possibly a way to make this work (and I have a super quick, not prod ready prototype) but it will take some thought because this may not even be the right fix. It's possible that this could/should be fixed at another level. We'll investigate to see what our options are.

In the meantime you can:
1. Add `../src/**/*.tsx` to your content array; OR
2. Add `../dist/index.cjs.js` to your content array; OR
3. Resolve the symlink yourself by doing this in your content array: `require("fs").realpathSync('./node_modules/react-tailwindcss/index.cjs.js')`

If we do figure out a proper way to fix this on our side it'll be some form of option 3. Have to make sure it works with globs and stuff though which will be the tricky part.

**Reply by tonytoth** [reply]
@thecrypticace thank you very much !! The third solution work like a charm.
We are trying to create a library and then use that library in a Next.js application, so we want to compile it in our Next.js so we avoid any conflicts and that's why I think this is a good solution for now.

**Reply by rix1** [reply]
Thanks so much @thecrypticace! I was on the verge of loosing it after debugging this issue for a couple of hours 😌 🧘 

I followed suggestion 3) and hot reloading now works flawlessly 🙌  Btw, I added glob support to your suggestion. Maybe this is something that could be covered by the official docs? (I'll submit a proof of concept PR).

For completeness, here's how I solved hot reloading:

**Context:**  I'm using Turborepo and have a shared UI package in `./packages/ui/`.

**apps/my-app/tailwind.config.js**
```js
const path = require('path')

module.exports = {
  content: [
    './pages/**/*.{ts,tsx}',
    './components/**/*.{ts,tsx}',
    // Hot reloading doens't work with symlinks, so we resolve the full path of the linked dependency
    // see https://github.com/tailwindlabs/tailwindcss/discussions/7717#discussioncomment-2310984
    path.join(require("fs").realpathSync('./node_modules/ui/'), 'src/**/*.{ts,tsx}'),
  ],
  theme: {
    extend: {},
  },
  presets: [require('config/tailwind.config')],
};

```

**apps/my-app/package.json**
```js
{
  "dependencies": {
    // Other deps...
    "ui": "workspace:*"
  },
}

```

**Reply by rix1** [reply]
@thecrypticace I added a PR here https://github.com/tailwindlabs/tailwindcss.com/pull/1336

**Reply by rix1** [reply]
Hmm... After a bit of testing I think writing the path out in full (option 1) is better. Like, Tailwind CLI works perfectly, but Prettier (of all things) does not like this configuration. I guess the reason being that VScode is running Prettier from root, which means the UI package have to be installed at root as well 🥴  In addition, the Editor UI started lagging (e.g. search), probably because Prettier now have to import and resolve symlinks(?) 

Anyways, I'll revert my suggestion and close PR https://github.com/tailwindlabs/tailwindcss.com/pull/1336

**Comment by lirbank** [comment]
Not sure if this is related but I found a solution to getting Tailwind JIT work with Storybook hot reload (dev mode), posted here: https://github.com/tailwindlabs/tailwindcss/discussions/6570#discussioncomment-2569691
